presto-main/src/main/java/io/prestosql/operator/LookupJoinOperator.java

```diff
@@ -41,587 +44,685 @@
+
+        public JoinTransformation(
+                List<Type> probeTypes,
+                List<Type> buildOutputTypes,
+                JoinType joinType,
+                LookupSourceFactory lookupSourceFactory,
+                JoinProbeFactory joinProbeFactory,
+                Runnable afterClose,
+                OptionalInt lookupJoinsCount,
+                HashGenerator hashGenerator,
+                PartitioningSpillerFactory partitioningSpillerFactory,
+                JoinStatisticsCounter statisticsCounter,
+                DriverYieldSignal yieldSignal,
+                SpillContext spillContext,
+                MemoryTrackingContext memoryTrackingContext)
+        {
+            this.probeTypes = ImmutableList.copyOf(requireNonNull(probeTypes, "probeTypes is null"));
 
-        if (spillInfoSnapshot.hasSpilled()) {
-            page = spillAndMaskSpilledPositions(page, spillInfoSnapshot.getSpillMask());
-            if (page.getPositionCount() == 0) {
-                return;
-            }
-        }
+            requireNonNull(joinType, "joinType is null");
+            // Cannot use switch case here, because javac will synthesize an inner class and cause IllegalAccessError
+            probeOnOuterSide = joinType == PROBE_OUTER || joinType == FULL_OUTER;
 
-        // create probe
-        inputPageSpillEpoch = spillInfoSnapshot.getSpillEpoch();
-        probe = joinProbeFactory.createJoinProbe(page);
+            this.joinProbeFactory = requireNonNull(joinProbeFactory, "joinProbeFactory is null");
+            this.afterClose = requireNonNull(afterClose, "afterClose is null");
+            this.lookupJoinsCount = requireNonNull(lookupJoinsCount, "lookupJoinsCount is null");
+            this.hashGenerator = requireNonNull(hashGenerator, "hashGenerator is null");
+            this.lookupSourceFactory = requireNonNull(lookupSourceFactory, "lookupSourceFactory is null");
+            this.partitioningSpillerFactory = requireNonNull(partitioningSpillerFactory, "partitioningSpillerFactory is null");
+            this.lookupSourceProviderFuture = lookupSourceFactory.createLookupSourceProvider();
+            this.statisticsCounter = requireNonNull(statisticsCounter, "statisticsCounter is null");
 
-        // initialize to invalid join position to force output code to advance the cursors
-        joinPosition = -1;
-    }
+            this.pageBuilder = new LookupJoinPageBuilder(buildOutputTypes);
+            this.yieldSignal = requireNonNull(yieldSignal, "yieldSignal is null");
+            this.memoryTrackingContext = requireNonNull(memoryTrackingContext, "memoryTrackingContext is null");
+            this.spillContext = requireNonNull(spillContext, "spillContext is null");
+        }
 
-    private boolean tryFetchLookupSourceProvider()
-    {
-        if (lookupSourceProvider == null) {
-            if (!lookupSourceProviderFuture.isDone()) {
-                return false;
+        public ListenableFuture<?> isBlocked()
```

Comment:
make this method private and put it under use, that is `WorkProcessor.TransformationState<WorkProcessor<Page>> process(@Nullable Page page)`